server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java

```diff
@@ -277,39 +277,45 @@ private static boolean isAggregationSort(BucketOrder order) {
         }
     }
 
-    @Override
-    protected Aggregator doCreateInternal(Aggregator parent, CardinalityUpperBound cardinality, Map<String, Object> metadata)
-        throws IOException {
-        BucketCountThresholds bucketCountThresholds = new BucketCountThresholds(this.bucketCountThresholds);
+    public static BucketCountThresholds adjustBucketCountThresholds(BucketCountThresholds bucketCountThresholds, BucketOrder order) {
```

Comment:
I've been thinking about this method. I don't like we invoke this twice. Maybe we can do this differently:

1. What if we add an adjustedBucketCountThresholds field to `TermsAggregationBuilder`
2. This new field only gets set by the `supportsParallelCollection(...)` method. The `supportsParallelCollection(...)` would make a copy with adjusted shard size. Step 2 is optional. If no parallel execution will happen this field remains `null`.
3. In the `innerBuild(...)` method we provide `adjustedBucketCountThresholds` if it is not null and otherwise just `bucketCountThresholds` to `TermsAggregatorFactory`.

I think this way the logic how to adjust shard size remains in one place. Do you think this is a better approach as well?

Reply:
I think in principle that we are optimizing for something that has little cost and it's not big deal that we call it twice and throw away the result in one case. It's not a heavy operation. It is odd but it sounds like the way to avoid it would add complexity that is not necessarily justified. I will play with this though to see if I can find a better solution.

Reply:
For me this more about cleaner code. I prefer that the logic that increases shard size is in one place.

> I will play with this though to see if I can find a better solution.

Cool, thanks!

Reply:
I've been going back and forth on this. technically, the logic is still in one place, but it gets called twice, and the builder throws away the result. I don't necessarily find it cleaner to introduce an additional field in the builder, which does not get serialized, and the logic to provide one or the other. I think it is simpler to just call the method from the outside and simulate what it will do to decide whether to enable concurrency or not.

I have considered moving this all to the builder, and one idea could be to reuse rewrite, as that surely gets called when the builder has all the fields set, but we should create a separate instance of the builder rather than modifying the existing one in place, and it feels weird that some validation would happen in the rewrite and possibly throw exception. For these reasons I discarded that option.

I even tried to refactor the whole thing, because the current adjusting logic is mixed with validation, and that does not feel right, but I ran into a bunch of problems that I did not feel should be solved in this PR, as it introduces risks. I experimented with the idea of moving the adjusting logic to the builder entirely, but then every time one of the affected values is set we need to re-do the adjustment, which is prone to errors and causes issues.

I am thinking that the current approach in the PR is not too bad, how strongly do you feel about trying to make it cleaner? I don't think we have great options.

Reply:
Thanks for exploring all the listed options. I agree with you that all the options have downsides. 
Let's stick to the current approach, since it isn't too bad.

Ideally the number of slices should be determined after the aggregation factories are constructed. The aggregation factories would then determine whether parallel collection is possible. I'm not sure whether this is possible and it will be a bigger change.

Reply:
Yes, what you describe is technically possible but it involves moving AggregationBuilder#build to be called before the index searcher is acquired. That in turn requires trimming down the context needed, because the current AggregationContext requires a searcher (although I do not think it is needed to build the aggregator factories.